File: 98.py
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        st = [(root, -float('inf'), float('inf'))]
        while len(st) != 0:
            node, lb, ub = st.pop()
            if node is None:
                continue
            if not (lb < node.val < ub):
                return False
            st.append((node.left, lb, node.val))
            st.append((node.right, node.val, ub))
        return True

File: sol/98/98.cpp
struct Solution {
    bool verify(TreeNode *node, long lb, long ub){
        if (node == nullptr){
            return true;
        }
        if (node->val <= lb || node->val >= ub) {
            return false;
        }
        return this->verify(node->left, lb, node->val) && this->verify(node->right, node->val, ub);
    }
    inline bool isValidBST(TreeNode* root) {
        return this->verify(root, -2147483649l, 2147483648l);
    }
};
